### ISSUE TITLE
rxjs mismatch

### ISSUE THREAD

**Issue by andreineculau** [angular/angular#11239]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
```

**Current behavior**

~~Failure to install `npm install @angular/http@2.0.0-rc.5` with npm v2.~~~
Failure to install `@angular/http@2.0.0-rc.5` with **npm v3** without getting warnings.

**Expected/desired behavior**

Succesfull install with no warnings e.g. UNMET PEER DEPENDENCY.

---

Looking at https://github.com/angular/angular/blob/master/CHANGELOG.md we can agree that the RC releases do not follow a strict semantic version i.e. there are breaking changes from rc.n to rc.n+1 . Yet, with that in mind, a module `@angular/http@2.0.0-rc.5` has a peerDependency on `^@angular/core@2.0.0-rc.5` basically saying that it will work with core@rc.6.

Yet they have *strict* dependencies on different versions of rxjs, something that was touched on in https://github.com/angular/angular/issues/10389.

To illustrate the problem, this is what happens with npm v3.10.3:

```
$ npm install @angular/core@^2.0.0-rc.5 @angular/common@^2.0.0-rc.5 @angular/platform-browser@^2.0.0-rc.5 rxjs@5.0.0-beta.6 zone.js@^0.6.17 @angular/http@2.0.0-rc.5
/Users/andrei/tmp/foo
├── @angular/common@2.0.0-rc.6
├── @angular/core@2.0.0-rc.6
├── @angular/http@2.0.0-rc.5
├── @angular/platform-browser@2.0.0-rc.6
├── UNMET PEER DEPENDENCY rxjs@5.0.0-beta.6
└── zone.js@0.6.17
```

Unfortunately the "UNMET PEER DEPENDENCY" is not a clear message, but basically it's a warning because `@angular/core@2.0.0-rc.6` expects "rxjs@5.0.0-beta.11" not "-beta.6".

I hope I made my case clear, but basically it boils down to these comments https://github.com/angular/angular/issues/10389#issuecomment-236307802 https://github.com/angular/angular/issues/10389#issuecomment-236331036 meaning that peerDependencies on `@angular` modules should also be strictly versioned.

---

UPDATED 2016-09-02 ~07:00 GMT by "removing" npm v2 references:

> I have discovered this issue because we actually use npm v2 (v3 was/is buggy, and we are waiting for a fix). The only difference is that npm v2 will fail with a message like below (much clearer than the npm v3 warning):

```
npm ERR! peerinvalid Peer @angular/http@2.0.0-rc.5 wants rxjs@5.0.0-beta.6
npm ERR! peerinvalid Peer @angular/core@2.0.0-rc.6 wants rxjs@5.0.0-beta.11
```

> Now, I have seen your replies around similar issues https://github.com/angular/angular/issues/10632 https://github.com/angular/angular/issues/10144
saying you don't support npm v2, but **the issue at hand is only surfaced by npm v2, and the root cause lies elsewhere IMO as the issue can be seen also via npm v3, except it is not blocking.**

**Comment by Hanfrey** [comment]
In the https://github.com/angular/angular/blob/master/CHANGELOG.md its mentioned that the core needs rxjs@5.0.0-beta.11

PEER-DEPENDENCY UPDATES

core: zone.js@0.6.17
core: rxjs@5.0.0-beta.11
core: reflect-metadata dependency is now optional when in AOT mode
compiler-cli: typescript@2.0.2

**Comment by JanNorden** [comment]
Note that the commandline is only installing rc.5 versions. It is the ^-dependecy that leads into the rc.6 world. The basic problem is mixing strict dependencies with compatible ones. Especially  since changing a strict dependency (as was done for core...-rc.6) is a breaking change.

**Comment by wesleycho** [comment]
I don't think Angular 2 officially supports npm 2 - there is a note in the documentation [here](https://angular.io/docs/ts/latest/guide/npm-packages.html) about using npm 3.

**Comment by andreineculau** [comment]
I have updated the description, so as to make it clear **this is not an issue related to npm v2**.

**Comment by IgorMinar** [comment]
@andreialecu the problem is that you are installing @angular/http@2.0.0-rc.5 which depends on rxjs beta.6 and @angular/core-2.0.0-rc.6 which dependson rxjs beta.11.

make sure you install http@2.0.0-rc.6 and rxjs@5.0.0-beta.11 and you'll be all set.

@JanNorden npm 2 is so broken that we strongly recommend that you migrate to v3 which is better (not great, but better).

**Comment by andreineculau** [comment]
@IgorMinar you got this *all wrong* and closed this issue without understanding it.

My goal is to install `@angular/http@2.0.0-rc.5`, which has its `package.json` defined with

```
peerDependencies:
   { rxjs: '5.0.0-beta.6',
     '@angular/core': '^2.0.0-rc.5',
     '@angular/platform-browser': '^2.0.0-rc.5' },
```

`@angular/core@^2.0.0-rc.5` resolves today to `@angular/core@2.0.0-rc.6`, and its `package.json` is defined with

```
peerDependencies: { rxjs: '5.0.0-beta.11', 'zone.js': '^0.6.17' },
```

thus the mismatch on rxjs!

---

Have a look at this conflictual `package.json` that copies the **exact** definitions of `peerDependencies` as per `npm info @angular/http@2.0.0-rc.5`, recursively:

```js
{
  "name": "foobaz",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
      "@angular/http": "2.0.0-rc.5",
      // main dependency, has 3 peer dependencies:

      "rxjs": "5.0.0-beta.6", // peer dep of @angular/http@2.0.0-rc.5

      "@angular/platform-browser": "^2.0.0-rc.5", // peer dep of @angular/http@2.0.0-rc.5
      // has 2 peer dependencies:
      "@angular/core": "^2.0.0-rc.6", // #A
      "@angular/common": "^2.0.0-rc.6",

      // "@angular/core": "^2.0.0-rc.5", // peer dep of @angular/http@2.0.0-rc.5, overridden by the stricter #A
      // has 2 peer dependencies:
      "zone.js": "^0.6.17",
      // "rxjs": "5.0.0-beta.11" // CONFLICTING DEP VERSION !!!
  }
}
```

Please have a second read of this issue, and reopen it. Thanks.

**Comment by IgorMinar** [comment]
you can only have one rxjs version in your project. http@...rc5 depends on
rxjs beta-6, core@...rc6 depends on rxjs beta-11. either downgrade core to
@...rc5 or upgrade http to rc6.

**Comment by andreineculau** [comment]
Correct @IgorMinar! Just that I did not come up with those conflictual dependencies. I have just followed the exact deps and their versions from your (angular's) package.json files.

It is http@2.0.0-rc.5 that incorrectly states that it is compatible with any core module above 2.0.0-rc.5 and below 2.1.0 (that is what ^2.0.0-rc.5 translates to), when in fact the angular team is only sure at release time that http@2.0.0-rc.5 is compatible with core@2.0.0-rc.5.

--
http://andreineculau.com
